src/parse.rs

```diff
@@ -985,41 +1003,57 @@ pub(crate) fn parse_decimal(s: &str, ix: usize) -> Option<(usize, usize)> {
         .map(|val| (end, val))
 }
 
-/// Attempts to parse an identifier between the specified opening and closing
-/// delimiters.  On success, returns `Some((id, skip))`, where `skip` is how much
-/// of the string was used.
+/// Attempts to parse an identifier, optionally followed by a relative number between the
+/// specified opening and closing delimiters.  On success, returns
+/// `Some((id, relative, skip))`, where `skip` is how much of the string was used.
 pub(crate) fn parse_id<'a>(
     s: &'a str,
     open: &'_ str,
     close: &'_ str,
     allow_relative: bool,
-) -> Option<(&'a str, usize)> {
+) -> Option<(&'a str, Option<isize>, usize)> {
     debug_assert!(!close.starts_with(is_id_char));
 
-    if !s.starts_with(open) {
+    if !s.starts_with(open) || s.len() <= open.len() + close.len() {
         return None;
     }
 
     let id_start = open.len();
     let mut iter = s[id_start..].char_indices().peekable();
-    let after_id = if allow_relative && iter.next_if(|(_, ch)| *ch == '-').is_some() {
-        iter.find(|(_, ch)| !ch.is_ascii_digit())
-    } else {
-        iter.find(|(_, ch)| !is_id_char(*ch))
-    };
+    let after_id = iter.find(|(_, ch)| !is_id_char(*ch));
+
     let id_len = match after_id.map(|(i, _)| i) {
-        Some(id_len) if s[id_start + id_len..].starts_with(close) => Some(id_len),
-        None if close.is_empty() => Some(s.len()),
-        _ => None,
+        Some(id_len) => id_len,
+        None if close.is_empty() => s.len(),
+        _ => 0,
     };
-    match id_len {
-        Some(0) => None,
-        Some(id_len) => {
-            let id_end = id_start + id_len;
-            Some((&s[id_start..id_end], id_end + close.len()))
+
+    let id_end = id_start + id_len;
+    if id_len > 0 && s[id_end..].starts_with(close) {
+        return Some((&s[id_start..id_end], None, id_end + close.len()));
+    } else if !allow_relative {
+        return None;
+    }
+    let relative_sign = s.as_bytes()[id_end];
+    if relative_sign == b'+' || relative_sign == b'-' {
+        if let Some((end, relative_amount)) = parse_decimal(&s, id_end + 1) {
+            if s[end..].starts_with(close) {
+                if relative_amount == 0 && id_len == 0 {
+                    return None;
+                }
+                let mut relative_amount_signed = relative_amount as isize;
+                if relative_sign == b'-' {
+                    relative_amount_signed = 0 - relative_amount_signed;
+                }
```

Comment:
Can we do without `mut`:

```suggestion
                let relative_amount_signed = if relative_sign == b'-' {
                    -(relative_amount as isize)
                } else {
                    relative_amount as isize
                };
                
```

Reply:
nice, yes we can :+1: